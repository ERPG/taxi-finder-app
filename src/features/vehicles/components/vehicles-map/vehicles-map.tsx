import { MapContainer, TileLayer, Marker, Popup, useMap } from "react-leaflet";
import "leaflet/dist/leaflet.css";
import L from "leaflet";
import { UnifiedVehicle } from "../../models/vehicle";
import { CarIcon } from "@freenow/wave";
import ReactDOMServer from "react-dom/server";
import { useCallback, useEffect, useMemo, useRef } from "react";
import { useVehicleSelection } from "../../contexts/select-vehicle.context";

export type VehiclesMapProps = {
    vehicles: UnifiedVehicle[] | undefined;
}

const AdjustZoomToFit = ({ vehicles }: { vehicles: UnifiedVehicle[] }) => {
    const map = useMap();
    const isFirstRender = useRef(true);

    useEffect(() => {
        if (vehicles.length > 0) {
            const bounds = L.latLngBounds(
                vehicles.map((vehicle) => [vehicle.coordinates[1], vehicle.coordinates[0]])
            );
            map.fitBounds(bounds, { padding: [50, 50] });
            isFirstRender.current = false;
        }
    }, [vehicles, map]);

    return null;
};

const MapWrapper = ({ children }: { children: React.ReactNode }) => {
    const defaultCenter: [number, number] = [53.5511, 9.9937];

    return (
        <MapContainer
            center={defaultCenter}
            zoom={12}
            style={{ height: "100%", width: "100%", minHeight: "392px" }}
            data-testid={"vehicle-map-container"}
            aria-label="Vehicle Map"
            keyboard={true}
        >
            <TileLayer
                url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
                attribution="&copy; OpenStreetMap contributors"
                data-testid={"vehicle-map-tile-layer"}
            />
            {children}
        </MapContainer>
    );
}

const VehiclesMarkers = ({ vehicles }: VehiclesMapProps) => {
    const ref = useRef<HTMLDivElement>(null);

    const { selectedVehicleId, setSelectedVehicleId } = useVehicleSelection();

    useEffect(() => {
        if (ref.current) {
            L.DomEvent.disableClickPropagation(ref.current);
        }
    });

    const createCustomIcon = useCallback((isSelected: boolean) => {
        const iconHTML = ReactDOMServer.renderToString(
            <div
                style={{
                    width: "40px",
                    height: "40px",
                    backgroundColor: isSelected ? "#003366" : "#0066cc",
                    borderRadius: "50%",
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "center",
                    boxShadow: "0 2px 5px rgba(0,0,0,0.3)",
                }}
            >
                <CarIcon size={20} color="white" />
            </div>
        );

        return L.divIcon({
            className: "custom-icon",
            html: iconHTML,
            iconSize: [40, 40],
            popupAnchor: [0, -20],
        });
    }, []);

    const markerElements = useMemo(() => {
        return vehicles?.map((vehicle) => (
            <div ref={ref}>
                <Marker
                    key={vehicle.id}
                    position={[vehicle.coordinates[1], vehicle.coordinates[0]]}
                    icon={createCustomIcon(vehicle.id === selectedVehicleId)}
                    eventHandlers={{
                        click: (e) => {
                            e.originalEvent.stopPropagation();
                            setSelectedVehicleId(vehicle.id)
                        },
                    }}
                >
                    <Popup>
                        <strong>{vehicle.licencePlate}</strong>
                        <br />
                        {vehicle.address || "No address"}
                    </Popup>
                </Marker>
            </div>
        ));
    }, [vehicles, selectedVehicleId, setSelectedVehicleId, createCustomIcon]);

    return <>{markerElements}</>;
};

const VehiclesMap = ({ vehicles }: VehiclesMapProps) => (
    <MapWrapper>
        {vehicles && vehicles.length > 0 && <AdjustZoomToFit vehicles={vehicles} />}
        {vehicles && <VehiclesMarkers vehicles={vehicles} />}
    </MapWrapper>
);

export default VehiclesMap;