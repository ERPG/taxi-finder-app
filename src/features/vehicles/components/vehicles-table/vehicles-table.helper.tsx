import { BatteryEmptyIcon, Battery10Icon, Battery25Icon, Battery50Icon, Battery75Icon, BatteryFullIcon } from "@freenow/wave";

export const getBatteryIcon = (fuel: number | null) => {
    if (fuel === null || fuel < 0) return null;
    if (fuel === 0) return <BatteryEmptyIcon size={18} color="red" />;
    if (fuel <= 10) return <Battery10Icon size={18} color="orange" />;
    if (fuel <= 25) return <Battery25Icon size={18} color="yellow" />;
    if (fuel <= 50) return <Battery50Icon size={18} color="green" />;
    if (fuel <= 75) return <Battery75Icon size={18} color="green" />;
    return <BatteryFullIcon size={18} color="darkgreen" />;
};