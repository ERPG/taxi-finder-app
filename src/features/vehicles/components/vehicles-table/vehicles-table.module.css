.table {
    width: 100%;
}

.paginationContainer {
    margin-top: 20px;
}