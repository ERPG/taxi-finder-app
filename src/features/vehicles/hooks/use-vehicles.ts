import { useQuery } from "@tanstack/react-query";
import { fetchAllVehicles } from "../services/vehicles-service";
import { UnifiedVehicle } from "../models/vehicle";
import { PaginatedResults } from "../../../shared/models/pagination";

export const useVehicles = (limit: number, page: number) => {
  return useQuery<PaginatedResults<UnifiedVehicle[]>, Error>({
    queryKey: ["vehicles", limit, page],
    queryFn: async () => await fetchAllVehicles(limit, page),
    staleTime: 1000 * 60 * 5,
  });
};