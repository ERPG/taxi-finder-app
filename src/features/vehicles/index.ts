import VehiclesFeature from "./vehicles";

export default VehiclesFeature;