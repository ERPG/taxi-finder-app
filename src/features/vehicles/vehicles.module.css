.container {
    display: flex;
    flex-direction: row;
    gap: 20px;
    height: 100vh;
    padding: 20px;
    box-sizing: border-box;
}

.mapContainer {
    flex: 1;
    height: 392px;
    width: 569px;
    padding: 10px;
}

.tableContainer {
    flex: 1;
    height: auto;
}

@media (max-width: 1200px) {
    .mapContainer {
        width: 320px;
        height: 392px;
    }
}

@media (max-width: 993px) {
    .container {
        flex-direction: column;
    }

    .mapContainer {
        width: 100%;
    }
}

@media (max-width: 768px) {
    .mapContainer {
        width: 100%;
        height: 60vh;
    }
}