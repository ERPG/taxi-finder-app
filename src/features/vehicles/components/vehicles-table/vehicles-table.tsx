import { Label, Pagination, Table, TableCell, TableHeaderCell, TableRow, TableRowSkeleton, TableSortableHeaderCell, Text, useSortBy } from "@freenow/wave";
import { EmojiHappyIcon, EmojiSadIcon } from '@freenow/wave';
import { PaginatedResults } from "../../../../shared/models/pagination";
import { UnifiedVehicle } from "../../models/vehicle";

import styles from './vehicles-table.module.css';
import { useCallback, useMemo } from "react";

import { orderBy } from "lodash";
import { getBatteryIcon } from "./vehicles-table.helper";
import { useVehicleSelection } from "../../contexts/select-vehicle.context";

export type VehiclesTableProps = {
    vehicles: PaginatedResults<UnifiedVehicle[]> | undefined;
    currentPage: number;
    onCurrentPageChange: (id: React.SetStateAction<number>) => void;
    isLoading: boolean;
}

const VehiclesTable = ({ vehicles, isLoading, currentPage, onCurrentPageChange }: VehiclesTableProps) => {
    const lastPage = Math.ceil((vehicles?.totalItems ?? 1) / (vehicles?.pageSize ?? 1));

    const { selectedVehicleId, setSelectedVehicleId } = useVehicleSelection();

    const { sortBy, setSortBy } = useSortBy('licencePlate', 'ASC');

    const handleSortingChange = useCallback(
        (field: string) => {
            console.log();
            setSortBy(field);
        },
        [setSortBy]
    );
    
    const direction = sortBy.direction?.toLowerCase() || 'asc';
    // @ts-expect-error Argument of type 'string' is not assignable to parameter of type 'Many<boolean | "asc" | "desc"> | undefined'.ts(2769)
    const sortedVehicles = orderBy(vehicles?.data, sortBy.field, direction);

    const rows = useMemo(() => {
        return sortedVehicles?.map((vehicle) => (
            <TableRow
                aria-selected={vehicle.id === selectedVehicleId}
                data-testid={"vehicle-row-" + vehicle.id}
                key={vehicle.id}
                onClick={() => setSelectedVehicleId(vehicle.id)}
                active={vehicle.id === selectedVehicleId}
                tabIndex={0}
                role="row"
                aria-label={`Vehicle ${vehicle.licencePlate}, ${vehicle.state}`}
                onKeyDown={(event) => {
                    if (event.key === "Enter" || event.key === " ") {
                        setSelectedVehicleId(vehicle.id);
                    }
                }}
            >
                <TableCell>{vehicle.type}</TableCell>
                <TableCell>{vehicle.licencePlate}</TableCell>
                <TableCell>{`${vehicle.coordinates[1]}, ${vehicle.coordinates[0]}`}</TableCell>
                <TableCell>{vehicle.address || "-"}</TableCell>
                <TableCell>
                    <Label 
                        variant={vehicle.state === "ACTIVE" ? "success" : "danger"}
                        aria-label={`Vehicle is ${vehicle.state}`} 
                    >
                        {vehicle.state === "ACTIVE" ? "Active" : "Inactive"}
                    </Label>
                </TableCell>
                <TableCell>
                    <div style={{ display: "flex", flexDirection: "row" }}>
                        {getBatteryIcon(vehicle.fuel)}
                        {vehicle.condition === "GOOD" ? (
                            <div>
                                <EmojiHappyIcon size={18} color="green" aria-label="Good Condition" />
                            </div>
                        ) : (
                            <EmojiSadIcon size={18} color="red" aria-label="Bad Condition" />
                        )}
                    </div>
                </TableCell>
            </TableRow>
        ));
    }, [sortedVehicles, selectedVehicleId, setSelectedVehicleId]);

    return (
        <div data-testid="vehicles-feature">
            <Table rowStyle="zebra" className={styles.table} data-testid="vehicles-table">
                <thead data-testid="vehicles-table-head" aria-label="Vehicles table">
                    <TableRow>
                        <TableHeaderCell scope="col">Type</TableHeaderCell>
                        <TableSortableHeaderCell
                            field="licencePlate"
                            active={sortBy.field === 'licencePlate'}
                            direction={sortBy.direction}
                            onSortChange={handleSortingChange}
                            aria-sort={sortBy.direction === 'ASC' ? 'ascending' : 'descending'}
                            role="columnheader"
                        >
                            Licence Plate
                        </TableSortableHeaderCell>
                        <TableHeaderCell scope="col">Coordinates</TableHeaderCell>
                        <TableHeaderCell scope="col">Address</TableHeaderCell>
                        <TableHeaderCell scope="col">State</TableHeaderCell>
                        <TableHeaderCell scope="col">Condition</TableHeaderCell>
                    </TableRow>
                </thead>
                <tbody>{isLoading ? Array.from({ length: vehicles?.pageSize ?? 1 }, (_, i) => i + 1).map(key => <TableRowSkeleton key={key} columns={6} />) : rows}</tbody>
            </Table>
            {!isLoading ? (
                <div className={styles.paginationContainer} data-testid="vehicles-pagination-container">
                    <Pagination
                        value={currentPage}
                        pageSize={vehicles?.pageSize ?? 1}
                        totalItems={vehicles?.totalItems ?? 1}
                        aria-label="Pagination controls"
                        ariaLabelPrevious="Prev"
                        ariaLabelNext="Next"
                        label={
                            <Text>
                                Page{' '}
                                <Text as="b" fontWeight="semibold">
                                    {currentPage} of {vehicles?.totalPages}
                                </Text>
                            </Text>
                        }
                        onNextPage={() => onCurrentPageChange(current => current + 1)}
                        onPrevPage={() => onCurrentPageChange(current => current - 1)}
                        onSkipForward={() => onCurrentPageChange(lastPage)}
                        onSkipBackward={() => onCurrentPageChange(1)}
                    />
                </div>) : null}
        </div>
    );
};

export default VehiclesTable;