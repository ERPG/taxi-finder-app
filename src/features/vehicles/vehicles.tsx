import { Suspense, useState } from "react";
import useErrorBoundary from "../../shared/components/error-boundary/error-boundary";
import VehiclesTable from "./components/vehicles-table/vehicles-table";
import VehiclesMap from "./components/vehicles-map/vehicles-map";
import { useVehicles } from "./hooks/use-vehicles";

import styles from "./vehicles.module.css";
import { PAGE_LIMIT } from "../../shared/utils/constants";
import ErrorFallback from "@/shared/components/error-fallback/error-fallback";
import { VehicleSelectionProvider } from "./contexts/select-vehicle.context";

const VehiclesFeature = () => {
    const [currentPage, setCurrentPage] = useState<number>(1);
    const { ErrorBoundaryWrapper } = useErrorBoundary();

    const { data: vehicles, isLoading, isError, refetch } = useVehicles(PAGE_LIMIT, currentPage);

    if (isLoading) return <p>Loading...</p>;
    if (isError) return <ErrorFallback refetch={refetch} />;

    const handlePageChange = (page: React.SetStateAction<number>) => {
        setCurrentPage(page);
    };

    return (
        <ErrorBoundaryWrapper>
            <Suspense fallback={<div>Loading vehicles...</div>}>
                <VehicleSelectionProvider>
                    <div className={styles.container}>
                        <div className={styles.mapContainer}>
                            <VehiclesMap vehicles={vehicles?.data} />
                        </div>
                        <div className={styles.tableContainer}>
                            <VehiclesTable
                                vehicles={vehicles}
                                currentPage={currentPage}
                                onCurrentPageChange={handlePageChange}
                                isLoading={isLoading}
                            />
                        </div>
                    </div>
                </VehicleSelectionProvider>
            </Suspense>
        </ErrorBoundaryWrapper>
    );
};

export default VehiclesFeature;
