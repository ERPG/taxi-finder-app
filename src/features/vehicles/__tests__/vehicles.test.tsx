import { fireEvent, render, screen } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import VehiclesFeature from "..";
import { QueryClientTestWrapper } from "../../../tests/utils/query-client";
import { mockPaginatedResults } from "../../../tests/__mocks__/vehicles";
import { VehiclesTableProps } from "../components/vehicles-table/vehicles-table";
import { VehiclesMapProps } from "../components/vehicles-map/vehicles-map";
import { PAGE_LIMIT } from "@/shared/utils/constants";
import { useVehicles } from "../hooks/use-vehicles";

vi.mock("../hooks/use-vehicles", () => ({
    useVehicles: vi.fn(),
}));

vi.mock("../components/vehicles-map/vehicles-map", () => ({
    default: ({ vehicles }: VehiclesMapProps) => (
        <div>
            Vehicles Map - results: {vehicles?.length}
        </div>
    ),
}));

vi.mock("../components/vehicles-table/vehicles-table", () => ({
    default: ({ vehicles, onCurrentPageChange, currentPage }: VehiclesTableProps) => (
        <div>
            Vehicles Table - results: {vehicles?.totalItems} - page: {currentPage}
            <button
                onClick={() => onCurrentPageChange(2)}
                data-testid="change-page-button"
            >
                Change Page
            </button>
        </div>
    ),
}));

describe('vehicles feature component suite', () => {
    beforeEach(() => {
        (useVehicles as ReturnType<typeof vi.fn>).mockReturnValue({
            data: mockPaginatedResults,
            isLoading: false,
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    const vehiclesFeatureComp = (
        <QueryClientTestWrapper>
            <VehiclesFeature />
        </QueryClientTestWrapper>
    )

    test("Should match snapshot", () => {
        const { container } = render(vehiclesFeatureComp);
        expect(container).toMatchSnapshot();
    });

    test("should call useVehicles with correct parameters", () => {
        (useVehicles as ReturnType<typeof vi.fn>).mockReturnValue({
            data: null,
            isLoading: true,
        });

        render(vehiclesFeatureComp);

        expect(useVehicles).toHaveBeenCalledWith(PAGE_LIMIT, 1);
    });

    test("should render loading state initially", () => {
        (useVehicles as ReturnType<typeof vi.fn>).mockReturnValue({
            data: null,
            isLoading: true,
        });

        render(vehiclesFeatureComp);

        expect(screen.getByText("Loading...")).toBeInTheDocument();
    });

    test("should render vehicles data in child components when loaded", () => {
        render(vehiclesFeatureComp);

        expect(screen.getByText("Vehicles Map - results: 1")).toBeInTheDocument();
        expect(screen.getByText("Vehicles Table - results: 1 - page: 1")).toBeInTheDocument();
    });

    test("should update currentPage when onCurrentPageChange is called", () => {
        render(vehiclesFeatureComp);

        const changePageButton = screen.getByTestId("change-page-button");
        fireEvent.click(changePageButton);

        expect(screen.getByText("Vehicles Table - results: 1 - page: 2")).toBeInTheDocument();
    });
})