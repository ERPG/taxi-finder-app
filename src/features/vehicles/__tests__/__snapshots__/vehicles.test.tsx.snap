// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vehicles feature component suite > Should match snapshot 1`] = `
<div>
  <div
    class="_container_db862b"
  >
    <div
      class="_mapContainer_db862b"
    >
      <div>
        Vehicles Map - results: 
        1
      </div>
    </div>
    <div
      class="_tableContainer_db862b"
    >
      <div>
        Vehicles Table - results: 
        1
         - page: 
        1
        <button
          data-testid="change-page-button"
        >
          Change Page
        </button>
      </div>
    </div>
  </div>
</div>
`;
